File: Solutions/Pancake_Sorting/Pancake_Sorting.js
var pancakeSort = function(a, len, arr=[], max=a.length) {
    while(max>1){
        flip(a.indexOf(max));
        max--;  
    }
    function flip(indx){
        if(indx){
            for(let q=0; q<indx/2; q++){
                [[a[q], a[indx-q]] = [a[indx-q], a[q]]];
            };
            arr.push(indx+1);
        }
        len=max-1;
        for(let q=0; q<len/2; q++){
            [[a[q], a[len-q]] = [a[len-q], a[q]]];
        };
        arr.push(max);
    }
    return arr;
};


/**
 * @param {number[]} A
 * @return {number[]}
 */
var pancakeSort = function(A) {
    let res = [];
    let curr = A.length;
    while (curr > 0) {
        const idx = A.indexOf(curr);
        if (idx != curr - 1) {
            res.push(idx + 1);
            flip(idx, A);
            res.push(curr);
            flip(curr - 1, A);
        }
        curr--;
    }
    return res;
};

const flip = (i, A) => {
    for(let j = 0; j <= Math.floor(i / 2); j++) {
        [A[i - j], A[j]] = [A[j], A[i - j]];
    }
};


//Fastest
/**
 * @param {number[]} A
 * @return {number[]}
 */

var pancakeSort = function (A) {
    const flip = (arr, i) => {
        let left=0
        let right =i
        while(left<right){
            temp = arr[right]
            arr[right]=arr[left]
            arr[left]=temp
            left++
            right--
        }
    }
    const findMax = (arr, i) => {
        let mInd = 0
        let isSorted = true, last = arr[0]
        for (let j = 0; j <= i; j++) {
            if (A[j] > A[mInd]) {
                mInd = j
            }
            if (isSorted && A[j] < last) isSorted = false
            last = A[j]
        }
        return [mInd, isSorted]
    }

    if (!A || A.length <=1 ) { return [] }
    let results = []
    let iterInd = A.length - 1

    // dont iterate over the zero index. It is sorted by the prior sort
    while (iterInd > 0) {
        let [maxInd, isSorted] = findMax(A, iterInd)
        // skip iterations if the items are sorted
        if(isSorted) return results
        // flip so the largest index is on top
        flip(A, maxInd)
        results.push(maxInd+1)
        // flip so the largest index is in place
        flip(A, iterInd)
        results.push(iterInd+1)

        iterInd--
    }
    return results
};

File: Solutions/Pancake_Sorting/Pancake_Sorting.cs
public class Solution {
    public IList<int> PancakeSort(int[] A) {
        var currVal = A.Length;
        var answer = new List<int>();
        while(currVal > 0) {
            var idx = Array.IndexOf(A, currVal);
            
            if (idx == currVal - 1) { // current value is in right place, go to next one
                currVal--;
                continue;
            }
            
            if (idx != 0) {  // if current value is not at index 0, flip it to first
                answer.Add(idx + 1);
                Flip(A, idx);
                //Console.WriteLine($"a: {string.Join(",", A)}");
            }
            answer.Add(currVal);  // flip current value to right index
            Flip(A, currVal - 1);
            //Console.WriteLine($"b: {string.Join(",", A)}");
            currVal--;
        }
        
        return answer;
    }
    
    private void Flip(int[] A, int kIndex) {
        var l = 0;
        var r = kIndex;
        while(l < r) {
            var tmp = A[l];
            A[l] = A[r];
            A[r] = tmp;
            l++;
            r--;
        }
    }
}

//Fastest
public class Solution {
    public IList<int> PancakeSort(int[] A) {
        // find max element
        var solution = new List<int>();
        solve(A, A.Length, solution);
        
        return solution;
        
    }
    
    public void solve(int[] A, int length, IList<int> solution) {
        /*foreach(var e in A){
            Console.WriteLine(e);
        }
        Console.WriteLine("foo");*/
        if(length == 1){
            return;
        }
        long max = Int32.MinValue;
        int i = 0;
        
        for(int j = 0; j<length; j++){
            if(A[j] > max){
                i = j;
                max = A[j];
            }
        }
        
        if(i != length-1){
        Array.Reverse(A, 0, i+1);
        Array.Reverse(A, 0, length);
        solution.Add(i+1);
        solution.Add(length);
        }
        
        solve(A, length-1, solution);
    }
}